geode-junit/src/main/java/org/apache/geode/test/junit/rules/ExecutorServiceRule.java

```diff
@@ -0,0 +1,361 @@
+ *   Future<Void> thread1 = asynchronousRule.runAsync(() -> {
+ *     try {
+ *       hangLatch.await();
+ *     } catch (InterruptedException e) {
+ *       throw new RuntimeException(e);
+ *     }
+ *   });
+ *
+ *   Future<Void> thread2 = asynchronousRule.runAsync(() -> {
+ *     try {
+ *       hangLatch.await();
+ *     } catch (InterruptedException e) {
+ *     throw new RuntimeException(e);
+ *     }
+ *   });
+ *
+ *   assertThatThrownBy(() -> thread1.get(1, MILLISECONDS)).isInstanceOf(TimeoutException.class);
+ * }
+ * </pre>
+ *
+ * <p>
+ * Example with awaitTermination enabled (awaits up to timeout for all submitted tasks to terminate
+ * after being interrupted). This causes the {@code Rule} to await termination of all threads during
+ * its tear down. This would only be interesting if you're trying to discover tasks that might not
+ * be responsive to interrupts such as blocking IO reads on Windows:
+ *
+ * <pre>
+ * private CountDownLatch hangLatch = new CountDownLatch(1);
+ *
+ * {@literal @}Rule
+ * public AsynchronousRule asynchronousRule = AsynchronousRule.builder().threadCount(10).awaitTermination(10, MILLISECONDS).build();
+ *
+ * {@literal @}Test
+ * public void doTest() throws Exception {
+ *   for (int i = 0; i < 10; i++) {
+ *     asynchronousRule.runAsync(() -> {
+ *       try {
+ *         hangLatch.await();
+ *       } catch (InterruptedException e) {
+ *         // do nothing
+ *       }
+ *     });
+ *   }
+ * }
+ * </pre>
+ */
+@SuppressWarnings("unused")
+public class ExecutorServiceRule extends SerializableExternalResource {
+
+  protected final int threadCount;
+  protected final boolean assertTasksAreDone;
+  protected final boolean awaitTermination;
+  protected final long awaitTerminationTimeout;
+  protected final TimeUnit awaitTerminationTimeUnit;
+
+  protected transient volatile ExecutorService executor;
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  protected ExecutorServiceRule(Builder builder) {
+    this.threadCount = builder.threadCount;
+    this.assertTasksAreDone = builder.assertTasksAreDone;
+    this.awaitTermination = builder.awaitTermination;
+    this.awaitTerminationTimeout = builder.awaitTerminationTimeout;
+    this.awaitTerminationTimeUnit = builder.awaitTerminationTimeUnit;
+  }
+
+  public ExecutorServiceRule() {
+    this.threadCount = 1;
+    this.assertTasksAreDone = false;
+    this.awaitTermination = false;
+    this.awaitTerminationTimeout = 0;
+    this.awaitTerminationTimeUnit = TimeUnit.NANOSECONDS;
+  }
+
+  @Override
+  public void before() {
+    if (threadCount > 1) {
+      executor = Executors.newFixedThreadPool(threadCount);
+    } else {
+      executor = Executors.newSingleThreadExecutor();
+    }
+  }
+
+  @Override
+  public void after() {
+    if (assertTasksAreDone) {
+      assertThat(executor.shutdownNow()).isEmpty();
+
+    } else {
+      executor.shutdownNow();
+    }
+    if (awaitTermination) {
```

Comment:
Shouldn't awaiting termination happen before we call executor.shutdownNow()?

Reply:
No, ExecutorService.awaitTermination is supposed to be invoked after calling shutdownNow. The only reason we would ever use this on ExecutorServiceRule is to find a task in some test that doesn't respond to interrupts.
```java
    /**
     * Blocks until all tasks have completed execution after a shutdown
     * request, or the timeout occurs, or the current thread is
     * interrupted, whichever happens first.
     *
     * @param timeout the maximum time to wait
     * @param unit the time unit of the timeout argument
     * @return {@code true} if this executor terminated and
     *         {@code false} if the timeout elapsed before termination
     * @throws InterruptedException if interrupted while waiting
     */
    boolean awaitTermination(long timeout, TimeUnit unit)
        throws InterruptedException;
```

Reply:
I like your interpretation better! I'm going to make some further changes to the Rule.